res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+  if (padMode === TraktorS2MK2.pad_modes.hotcue) {
+    for (var i = 1; i <= 4; i++) {
+      TraktorS2MK2.pad_connections[group].push(
+        engine.makeConnection(group, "hotcue_" + i + "_enabled", TraktorS2MK2.outputHotcueCallback));          
+      TraktorS2MK2.pad_connections[group].push(
+        engine.makeConnection(group, "hotcue_" + i + "_color", TraktorS2MK2.outputHotcueCallback));
+    }
+  } else if (padMode === TraktorS2MK2.pad_modes.intro_outro) {
+    for (var i = 1; i <= 4; i++) {
+      // This function to create callback functions is needed so the loop index variable
+      // i does not get captured in a closure within the callback.
+      var makeIntroOutroCallback = function(padNumber) {
+        return function(value, group, control) {
+          if (value > 0) {
+            TraktorS2MK2.sendPadColor(group, padNumber, intro_outro_colors[padNumber-1]);
+          } else {
+            TraktorS2MK2.sendPadColor(group, padNumber, {red: 0, green: 0, blue: 0});
+          }
+        };
+      };
+      TraktorS2MK2.pad_connections[group].push(engine.makeConnection(
+        group, intro_outro_keys[i-1] + "_enabled", makeIntroOutroCallback(i)));
+    }
+  } else if (padMode === TraktorS2MK2.pad_modes.sampler) {
+    for (var i = 1; i <= 4; i++) {
+      var makeSamplerCallback = function(deckGroup, padNumber) {
+        var samplerNumber = deckGroup === "[Channel1]" ? padNumber : padNumber + 4;
+        var samplerGroup = "[Sampler" + samplerNumber + "]";
+        return function(value, group, control) {
+          if (engine.getValue(samplerGroup, "track_loaded")) {
+            if (engine.getValue(samplerGroup, "play") === 1) {
+              if (engine.getValue(samplerGroup, "repeat") === 1) {
+                TraktorS2MK2.sendPadColor(deckGroup, padNumber, {red: 255, green: 255, blue: 0});
+              } else {
+                TraktorS2MK2.sendPadColor(deckGroup, padNumber, {red: 0xAF, green: 0x00, blue: 0xCC});
+              }
+            } else {
+              TraktorS2MK2.sendPadColor(deckGroup, padNumber, {red: 255, green: 255, blue: 255});
+            }
+          } else {
+            TraktorS2MK2.sendPadColor(deckGroup, padNumber, {red: 0, green: 0, blue: 0});
+          }
+        };
+      }
+      
+      var sNumber = group === "[Channel1]" ? i : i + 4;
+      var sGroup = "[Sampler" + sNumber + "]";
+      TraktorS2MK2.pad_connections[group].push(engine.makeConnection(
+        sGroup, "track_loaded", makeSamplerCallback(group, i)));
+      TraktorS2MK2.pad_connections[group].push(engine.makeConnection(
+        sGroup, "play", makeSamplerCallback(group, i)));
+      TraktorS2MK2.pad_connections[group].push(engine.makeConnection(
+        sGroup, "repeat", makeSamplerCallback(group, i)));
+    }
+  }
+  
+  TraktorS2MK2.pad_connections[group].forEach(function(connection) {
+    connection.trigger();
+  });
+  
+  TraktorS2MK2.current_pad_mode[group] = padMode;
+}
+
+TraktorS2MK2.pad = function(field) {
```

Comment:
this function is quite long -- can you split it into three, one that handles samplers, one that handles hotcues, and one that decides which to call?

Reply:
good idea